hnn_core/optimization/general_optimization.py

```diff
@@ -0,0 +1,409 @@
+"""Parameter optimization."""
+
+# Authors: Carolina Fernandez <cxf418@miami.edu>
+#          Nick Tolley <nicholas_tolley@brown.edu>
+#          Ryan Thorpe <ryan_thorpe@brown.edu>
+#          Mainak Jas <mjas@mgh.harvard.edu>
+
+import numpy as np
+
+from .metrics import _rmse_evoked
+
+from skopt import gp_minimize
+from scipy.optimize import fmin_cobyla
+
+
+class Optimizer:
+    def __init__(self, net, tstop, constraints, set_params, solver='bayesian',
+                 obj_fun='dipole_rmse', scale_factor=1.,
+                 smooth_window_len=None, max_iter=200):
+        """Parameter optimization.
+
+        Parameters
+        ----------
+        net : Network
+            The network object.
+        tstop : float
+            The simulated dipole's duration.
+        constraints : dict
+            The user-defined constraints.
+        set_params : func
+            User-defined function that sets parameters in network drives.
+        solver : string
+            The optimizer, 'bayesian' or 'cobyla'.
+        obj_fun : string
+            The objective function to be minimized.
+        scale_factor : float, optional
+            The dipole scale factor. The default is 1.
+        smooth_window_len : float, optional
+            The smooth window length. The default is None.
```

Comment:
In the future, we should probably try to find a way to confine these parameters to the `_rmse_evoked()` metric function because they're not a general property of the Optimizer.